CodingStandard.md

```diff
@@ -0,0 +1,1425 @@
+    foo_explicit(p);
+
+    // no error, explicit call to converting constructor
+    foo_explicit(Explicit{p});
+
+    // error: no matching function for call to 'bar_implicit'
+    bar_implicit(x, y);
+
+    // no error, implicit conversion to Point
+    bar_implicit(Implicit(x, y));
+
+    // error: no matching function for call to 'bar_explicit'
+    bar_explicit(Explicit(x, y));
+
+    // no error, explicit conversion to Point
+    bar_explicit(static_cast< Point >(Explicit(x, y)));
+
+    int a = 1; int b = 2;
+
+    // no error
+    Implicit i1{a, b};
+
+    // no error
+    Explicit e1{a, b};
+
+    // no error
+    Implicit i = {a, b};
+
+    // error: chosen constructor is explicit in copy-initialization
+    Explicit e = {a, b};
+
+    // no error
+    Explicit e = Explicit{a, b};
+
+    // no error
+    Explicit e = Explicit{{a, b}};
+}
+```
+
+Another way to avoid unintentional implicit or narrowing conversions is to
+initialize variables using curly braces instead of parentheses.  This helps
+expose unintended narrowing conversions of built-in types in some cases, but
+not all.
+
+```c++
+int32_t foo()
+{
+    return 200;
+}
+int8_t bar(int8_t x)
+{
+    return x + 100;  // undefined behavior if x > 27
+}
+int main()
+{
+    int32_t a = 100000;
+    int8_t  b{a};        // error: narrowing conversion of 'a' inside { }
+    int8_t  b = {a};     // error: narrowing conversion of 'a' inside { }
+    int8_t  b = a;       // silent narrowing of a
+
+    int32_t x = 10000000000;    // warning: implicit conversion changes value
+    int32_t x = {10000000000};  // error: constant expression cannot be narrowed
+
+    int32_t y = foo();
+    int32_t z = bar(y);       // silent narrowing of y inside bar()
+    return z;                 // returns ??
+}
+```
+
+Brace initializers should also be used to avoid narrowing conversions in
+constructor initializer lists and when calling an object's constructor.
+
+```c++
+struct Point
+{
+   Point(int x, int y) : x_{x}, y_{y} {}
+   int x_;
+   int y_;
+};
+int main()
+{
+    double x = 1.25; double y = 2.25;
+
+    // no error, despite narrowing double to int
+    Point p1(x, y);
+
+    // error: non-constant-expression cannot be narrowed in initializer list
+    Point p2{x, y};
+
+    // no error, despite narrowing double to int
+    Point p3 = Point(x, y);
+
+    // error: type 'double' cannot be narrowed to 'int' in initializer list
+    Point p4 = Point{x, y};
+}
+```
+
+**Software Exceptions and Exception Handling**
+
+<u>Summary</u>
+
+- Do not use the C++ exception standard library
+- Do not use exception-related keywords `try`, `catch`, `throw`, `noexcept`
+- Code is compiled using `-fno-exceptions` to disable exceptions
+- Define interfaces with narrow contracts and use assertions to ensure they
+  are fulfilled
+- Check all return values or explicitly discard them with a cast to `void`
+- Use `[[nodiscard]]` to ensure that return values are always checked or
+  explicitly discarded
```

Comment:
With embedded systems this is commonplace. But I think we need to define an explict uniform system of error codes, or use the ROS error codes.

Reply:
If ROS already has a set of defined error codes, I vote for using them so we don't have to think.